### PR TITLE
Use TTS progress listener status in Card Viewer auto-advance delay

### LINKED ISSUE #6087: Feature Request: working combination of auto-advance and text to speech
###### Reproduction Steps

1. Enable auto-answer cards
2. Enable TTS
3. Review a selection of cards

###### Expected Result

The review should auto-advance with TTS working correctly and the auto-advance happening according to the "time for TTS to speak" plus the configured auto-advance time

###### Actual Result

According to user feedback TTS and auto-advance do not work together, and it appears that the related code is indeed not connected.

A viable solution would be to plumb a TTS UtteranceProgressListener through from AbstractFlashcardViewer so that auto-advance could function, and when the listener onDone was called and the TTS queue was empty AbstractFlashcardViewer could apply auto-advance logic

https://github.com/ankidroid/Anki-Android/blob/8c90adb685b5970356173325d2febcf05ada0cae/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java#L2135-L2145

and connected here somehow

https://github.com/ankidroid/Anki-Android/blob/8c90adb685b5970356173325d2febcf05ada0cae/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java#L1876-L1883

###### Debug info
Refer to the [support page](https://ankidroid.org/docs/help.html) if you are unsure where to get the "debug info".

###### Research
*Enter an [ x ] character to confirm the points below:*

[  ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[  ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[  ] I have searched for similar existing issues here and on the user forum

### PR DESCRIPTION
**Pull request by jianersswlq**
## Purpose / Description
Feature Request: working combination of auto-advance and text to speech

## Fixes
Fixes #6087

## Approach
Initialize TTS in AbstractFlashCardViewer.java and set UtteranceProgressListener in this class.

## How Has This Been Tested?

Enable auto-answer cards
Enable TTS
Review a selection of cards

Android version 2.10alpha

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/ReadText.java [hunk 1]
```diff
@@ -1,369 +1,329 @@
...
+ ****************************************************************************************/
+
+package com.ichi2.anki;
+
+import android.content.Context;
+import android.content.res.Resources;
+import android.net.Uri;
+import android.os.Handler;
+import android.speech.tts.TextToSpeech;
+import android.speech.tts.UtteranceProgressListener;
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/ReadText.java:24]: Codacy Issue found: [Avoid unused imports such as 'android.speech.tts.UtteranceProgressListener'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5540429)

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Complexity decreasing per file
==============================
+ AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5540429&bid=18163352)

File: AnkiDroid/src/main/java/com/ichi2/anki/ReadText.java [hunk 2]
```diff
@@ -1,369 +1,329 @@
...
+import android.view.View;
+
+import android.view.WindowManager;
+import android.widget.Toast;
+
+import com.afollestad.materialdialogs.MaterialDialog;
+import com.google.android.material.snackbar.Snackbar;
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/ReadText.java:31]: Codacy Issue found: [Avoid unused imports such as 'com.google.android.material.snackbar.Snackbar'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5540429)

File: AnkiDroid/src/main/java/com/ichi2/anki/ReadText.java [hunk 3]
```diff
@@ -1,369 +1,329 @@
...
+ *                                                                                      *
+ * You should have received a copy of the GNU General Public License along with         *
+ * this program.  If not, see <http://www.gnu.org/licenses/>.                           *
+ ****************************************************************************************/
+
+package com.ichi2.anki;
+
+import android.content.Context;
+import android.content.res.Resources;
+import android.net.Uri;
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/ReadText.java:21]: Codacy Issue found: [Avoid unused imports such as 'android.net.Uri'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5540429)

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 4]
```diff
@@ -1,3113 +1,3174 @@
...
+                                    v -> openTtsHelpUrl(helpUrl), ankiActivity.findViewById(R.id.root_layout),
+                                    new Snackbar.Callback());
+                        }
+                        @Override
+                        public void onStart(String arg0) {
+                            // no nothing
+                        }
+                    });
+                } else {
+                    Toast.makeText(ReadText.getmReviewer().get(), ReadText.getmReviewer().get().getString(R.string.no_tts_available_message), Toast.LENGTH_LONG).show();
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:915]: Can be shortened to: `UIUtils.showThemedToast(context, context.getString(R.string.no_tts_available_message), false);`

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 5]
```diff
@@ -1,3113 +1,3174 @@
...
-
-
-    /**
-     * Clean up the correct answer text, so it can be used for the comparison with the typed text
-     *
-     * @param answer The content of the field the text typed by the user is compared to.
-     * @return The correct answer text, with actual HTML and media references removed, and HTML entities unescaped.
-     */
-    protected String cleanCorrectAnswer(String answer) {
-        return TypedAnswer.cleanCorrectAnswer(answer);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1895]: This change has reverted some changes in `AbstractFlashCardViewer`, could you undo this so the commit only contains changes that you've made which are required for the feature.

Comment (jianersswlq) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1895]: Is it because the version I fork is not the latest?
How about creating a new pull request for that?

Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1895]: It's fine to request a pull request from an earlier branch. I don't quite see how you don't have merge conflicts since you're based off master, but moving to a branch back in time. It's a strange situation to be in.

I'd probably have a google to understand how I got into the situation and see what's advised.

Alternately: shelve to keep the changes safe, then: revert and reapply, but it's up to you.

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1895]: I'd go a little more firm: Always make a PR of current master, like master as of that moment, start fully up to date, then start making changes. Do not change whitespace unnecessarily for sure. I'm excited for this change though! We have had a few people asking for it, and my reply has been "please make a PR", now we have a PR :-). I'm sure we can get past the "PR mechanics" phase and get this in

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5540429&bid=18163352)

**Comment by jianersswlq** [comment]
I modify these two .java from latest version now. Could it be reviewed now?

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 6]
```diff
@@ -128,11 +123,47 @@
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
+import androidx.annotation.CheckResult;
+import androidx.annotation.IdRes;
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+import androidx.appcompat.app.ActionBar;
+import androidx.core.content.ContextCompat;
+import androidx.core.view.GestureDetectorCompat;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:132]: This diff looks a lot cleaner but I don't understand all the motion around imports - your proposed change does not affect these imports at all, or the expansion of the ViewerCommand ones below. Can you please revert these changes and just restrict the PR to only the exact things necessary to implement the feature?

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 7]
```diff
@@ -2726,7 +2837,8 @@ private void drawFlag() {
     }
 
 
-    protected @FlagDef int getFlagToDisplay() {
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:2729]: This change is unrelated to your PR - please revert

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 8]
```diff
@@ -2682,7 +2717,83 @@ public void playVideo(String path) {
...
+        }
+    }
+    public void showAnswerTask(){
+        long delay = mWaitAnswerSecond * 1000;
+        if (delay > 0) {
+            mTimeoutHandler.postDelayed(mShowAnswerTask, delay);
+        }
+    }
     /** Callback for when TTS has been initialized. */
+    public void initializeTts(Context context) {
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:2734]: I'm concerned in general by the motion of code into AbstractFlashcardViewer, that object is huge, and TTS was already separated from it nicely

I'm looking for any solution that involves keeping the initializeTts out of AbstractFlashcardViewer

It seems like the real change is the addition of this

```java

                            if(ReadText.getmQuestionAnswer()==Sound.SOUNDS_QUESTION){
                                showAnswerTask();
                            }else if(ReadText.getmQuestionAnswer()==Sound.SOUNDS_ANSWER){
                                showQuestionTask();
                            }
```

Perhaps as a different structure we could have a ReadTextListener interface inside ReadText that just had one method `onDone`

Have AbstractFlashcardViewer implement it, with the code above as the implementatio, add the listener argument as a parameter to initializeTts and as you setOnUtteranceProgressListener you bind them together by having the onDone + empty queue logic call ReadTextListener.onDone()

Then we can keep the TTS stuff in ReadText and not increase AbstractFlashcardViewer size much except the actual postDelayed logic

**Comment by jianersswlq** [comment]
OK, I will think about that and try.

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 9]
```diff
@@ -1795,6 +1795,21 @@ public void run() {
...
+                    mTimeoutHandler.postDelayed(mShowAnswerTask, delay);
+                }
+            }else if(ReadText.getmQuestionAnswer()==Sound.SOUNDS_ANSWER){
+                long delay = mWaitQuestionSecond * 1000;
+                if (delay > 0) {
+                    mTimeoutHandler.postDelayed(mShowQuestionTask, delay);
+                }
+            }
+        }
+    }
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1812]: Just adding whitespace according to our coding standards, logic seems fine does it work for you like this in your testing?

```suggestion
    class ReadTextListener implements ReadText.ReadTextListener {
        public void onDone() {
            if (ReadText.getmQuestionAnswer() == Sound.SOUNDS_QUESTION) {
                long delay = mWaitAnswerSecond * 1000;
                if (delay > 0) {
                    mTimeoutHandler.postDelayed(mShowAnswerTask, delay);
                }
            } else if (ReadText.getmQuestionAnswer() == Sound.SOUNDS_ANSWER) {
                long delay = mWaitQuestionSecond * 1000;
                if (delay > 0) {
                    mTimeoutHandler.postDelayed(mShowQuestionTask, delay);
                }
            }
        }
    }
```

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 10]
```diff
@@ -1877,7 +1892,9 @@ protected void displayCardQuestion() {
             long delay = mWaitAnswerSecond * 1000 + mUseTimerDynamicMS;
             if (delay > 0) {
                 mTimeoutHandler.removeCallbacks(mShowAnswerTask);
-                mTimeoutHandler.postDelayed(mShowAnswerTask, delay);
+                if(!mSpeakText){
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1895]: Same as above - whitespace changes only to match project standards

```suggestion
                if (!mSpeakText) {
```

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 11]
```diff
@@ -1954,7 +1971,9 @@ protected void displayCardAnswer() {
             long delay = mWaitQuestionSecond * 1000 + mUseTimerDynamicMS;
             if (delay > 0) {
                 mTimeoutHandler.removeCallbacks(mShowQuestionTask);
-                mTimeoutHandler.postDelayed(mShowQuestionTask, delay);
+                if(!mSpeakText){
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1974]: Same as above - whitespace changes only to match project standards

```suggestion
                if (!mSpeakText) {
```

**Comment by jianersswlq** [comment]
Thank you for the format. I think it is better and it works in my test now. However, I don't understand why there are some checks provided by github failed.

**Comment by mikehardy** [comment]
Well, the unit + lint tests that are blockers all passed, and the emulator tests are rolling green now, looks good to me? I think just the way the branch was managed github can figure out the parentage so it won't allow a rebase merge but a squash merge will work. That's fine by me. 

At this point we're down to the part of the pull request review where it's a matter of really looking at the core of it again and making sure it all makes sense

Last pass through when I did that it did seem to. I want to look at it one more time and maybe David has thoughts, we'll see pretty quickly

File: AnkiDroid/src/main/java/com/ichi2/anki/ReadText.java [hunk 12]
```diff
@@ -272,7 +275,7 @@ private static boolean isLanguageAvailable(String localeCode) {
                 TextToSpeech.LANG_AVAILABLE;
     }
 
-    public static void initializeTts(Context context) {
+    public static void initializeTts(Context context, ReadTextListener listener) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/ReadText.java:278]: Could you mark `listener` as non-null here to help future developers avoid programming errors:

`@NonNull ReadTextListener listener)`

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 13]
```diff
@@ -1795,6 +1795,21 @@ public void run() {
         }
     };
 
+    class ReadTextListener implements ReadText.ReadTextListener {
+        public void onDone() {
+            if (ReadText.getmQuestionAnswer() == Sound.SOUNDS_QUESTION) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1800]: Can we perform an early return here if `mUseTimer` is false?

Comment (jianersswlq) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1800]: Dose it mean we should check if(mUseTimer) here?

Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1803]: Yeah! something along the lines of

```java
if (!mUseTimer) {
    return;
}
```

Comment (jianersswlq) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1803]: I have already done. See the latest commit.

**Comment by mikehardy** [comment]
Oh sweet! I'm excited about this feature - I'll scan through it but if David likes it is ready I'm sure, then it'll go in the queue for the next alpha which I should kick out this morning, so you can see / play with it in an official version. As I scan it I'll see if I can pick it to 2.10.3 as well - it might be small enough to do so. Thanks

**Comment by mikehardy** [comment]
Going out now as 2.10.3beta3 and 2.11alpha6

**Comment by SergeyLeonov999** [comment]
Спасибо огромное за решение проблемы с TTS =продвижением., Остался ещё один маленький шаг и оно будет идеальным.
У меня есть карточки в которых заполнено только поле question, карточка Answer- пустая->. в поле ANSWER нет ни одного слова, TTS не запускается , отсчет задержки после TTS, не включается, так как  карточка ANSWER пустая. Карточка и колода останавливается. NO WORD IN ANSWER CARD (NO TTS) <-> (NO TTS ) no delay time after ANSWER CARD.

**Comment by mikehardy** [comment]
Hello! with the 2.14 release we started accepting donations with the stated goal of accepting money from users (the users' way of saying "Thanks!") to support maintenance and development of AnkiDroid.

We are looking through PRs merged in releases going back to release 2.10 (where we have good record keeping of what's in releases) to let our contributors know.

If you'd like to submit an expense related to your work here, to share in a portion of our hopefully-sustainable monthly disbursement, please read here and do so in the next couple of days:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process

This is our first time handling donations and attempting to compensate anyone with these processes, so please be flexible and patient with us

Thank you for your contribution to AnkiDroid!
